core/src/main/java/org/mule/runtime/core/execution/interception/InterceptorMessageProcessor.java

```diff
@@ -0,0 +1,207 @@
+ * <p/>
+ * It relies on {@link org.mule.runtime.api.meta.AnnotatedObject} to get the information for intercepting.
+ * In case of ObjectFactories it will be populated with the intercepted annotations.
+ *
+ * @since 4.0
+ */
+public class InterceptorMessageProcessor extends AbstractAnnotatedObject implements Processor, FlowConstructAware,
+    MuleContextAware, Lifecycle, Interceptor {
+
+  private transient Logger logger = LoggerFactory.getLogger(InterceptorMessageProcessor.class);
+
+  @Inject
+  private MuleContext muleContext;
+  private FlowConstruct flowConstruct;
+  private Processor intercepted;
+  private Processor next;
+
+  /**
+   * Sets the {@link Processor} being intercepted.
+   *
+   * @param intercepted {@link Processor} being intercepted.
+   */
+  public void setIntercepted(Processor intercepted) {
+    this.intercepted = intercepted;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public void setMuleContext(MuleContext muleContext) {
+    this.muleContext = muleContext;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void setListener(Processor listener) {
+    this.next = listener;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void initialise() throws InitialisationException {
+    if (intercepted == null) {
+      throw new InitialisationException(createStaticMessage("No processor has been set to be intercepted"), this);
+    }
+    if (this.getAnnotations().get(ANNOTATION_NAME) == null) {
+      throw new InitialisationException(createStaticMessage("Intercepted processor: " + intercepted.getClass().getName()
+          + " but no annotations where defined for it"), this);
+    }
+    if (intercepted instanceof FlowConstructAware) {
+      ((FlowConstructAware) intercepted).setFlowConstruct(flowConstruct);
+    }
+    initialiseIfNeeded(intercepted, true, muleContext);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void start() throws MuleException {
+    startIfNeeded(intercepted);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void stop() throws MuleException {
+    stopIfNeeded(intercepted);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void dispose() {
+    disposeIfNeeded(intercepted, logger);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void setFlowConstruct(FlowConstruct flowConstruct) {
+    this.flowConstruct = flowConstruct;
+  }
+
+  /**
+   * @return the {@link ProcessingType} of the intercepted {@link Processor}.
+   */
+  @Override
+  public ProcessingType getProcessingType() {
+    return intercepted.getProcessingType();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Event process(Event event) throws MuleException {
+    final ComponentIdentifier componentIdentifier = (ComponentIdentifier) this.getAnnotations().get(ANNOTATION_NAME);
+    Map<String, Object> parameters = ((Map) this.getAnnotations().getOrDefault(ANNOTATION_PARAMETERS, emptyMap()));
```

Comment:
is it correct to assume that this class is only intended to work with non SDK MPs? Otherwise, this won't interop well.